test/test_torch.py

```diff
@@ -2466,39 +2466,179 @@ def _test_rpow(self, cast):
+        nparr = np.array(ints, dtype=np.int64)
+
+        if device == 'cuda' and not TEST_WITH_ROCM:
+            with self.assertRaises(AssertionError):
+                # This is a check that pow CUDA implementation is
+                # incompatible with Numpy:
+                # pow CUDA  4 ^ 0.5 = 1
+                # numpy pow 4 ^ 0.5 = 2
+                # This check must be deleted after pow CUDA is fixed
+                for pow in floats:
+                    expected = np.power(nparr, pow).astype(np.int64)
+
+                    actual = tensor.pow(pow)
+                    self.assertEqual(expected, actual.cpu().numpy())
+
+                    actual = tensor.clone()
+                    actual2 = actual.pow_(pow)
+                    self.assertEqual(expected, actual.cpu().numpy())
+                    self.assertEqual(expected, actual2.cpu().numpy())
+
+                    actual = torch.pow(tensor, pow)
+                    self.assertEqual(expected, actual.cpu().numpy())
+
+                    actual2 = torch.pow(tensor, pow, out=actual)
+                    self.assertEqual(expected, actual.cpu().numpy())
+                    self.assertEqual(expected, actual2.cpu().numpy())
+        else:
+            for pow in floats:
+                expected = np.power(nparr, pow).astype(np.int64)
+
+                actual = tensor.pow(pow)
+                self.assertEqual(expected, actual.cpu().numpy())
+
+                actual = tensor.clone()
+                actual2 = actual.pow_(pow)
+                self.assertEqual(expected, actual.cpu().numpy())
+                self.assertEqual(expected, actual2.cpu().numpy())
+
+                actual = torch.pow(tensor, pow)
+                self.assertEqual(expected, actual.cpu().numpy())
+
+                actual2 = torch.pow(tensor, pow, out=actual)
+                self.assertEqual(expected, actual.cpu().numpy())
+                self.assertEqual(expected, actual2.cpu().numpy())
+
+    @torchtest.for_all_device_types()
+    @unittest.skipIf(not TEST_NUMPY, 'Numpy not found')
+    def test_float_scalar_pow_float_tensor(self, device):
+        def assertEqual(x, y):
+            self.assertEqual(np.nan_to_num(x), np.nan_to_num(y), allow_inf=True)
+        floats = [2.0, -3 / 2, -1.0, -1 / 2, -1 / 3, 0.0,
+                  1 / 3, 1 / 2, 1.0, 3 / 2, 2.0]
+
+        tensor = torch.tensor(floats, dtype=torch.float32, device=device)
+        nparr = np.array(floats, dtype=np.float32)
+        for base in floats:
+            expected = np.power(base, nparr).astype(np.float32)
+
+            actual = torch.pow(base, tensor)
+            assertEqual(expected, actual.cpu().numpy())
+
+            actual2 = torch.pow(base, tensor, out=actual)
+            assertEqual(expected, actual.cpu().numpy())
+            assertEqual(expected, actual2.cpu().numpy())
+
+    @torchtest.for_all_device_types()
+    @unittest.skipIf(not TEST_NUMPY, 'Numpy not found')
+    def test_tensor_pow_tensor(self, dev):
```

Comment:
I doubt that it is necessary, because this generic `test_pow` should cover all conditions like:

- if base is tensor and exp is tensor then we need a loop to rotate exp tensor...
- if base is scalar and exp is tensor then we need a loop thru all base values...
- if base is tensor and exp is scalar then we need a loop thru all exp values and we need to check that cpu/numpy do not cast exp to tensor.dtype but current cuda implementation does...
- etc...

Reply:
I am pointing out that there is a lot of code duplication between test_int_pow, test_long_tensor_pow_float,
and test_tensor_pow_tensor. At the very least, we should aim to make code easier to read by not duplicating it.

Reply:
addressed